Agda cubical

ryanorendorff opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
 
Adds the cubical type theory library to nixpkgs. I asked the maintainers and @Saizan gave approval to add this library to nixpkgs.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) (checked for any tests, there do not appear to be).
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (all agda files compile)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date (I cannot find a place in the documentation for Agda things)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @jwiegley @laMudri

ryanorendorff committed:
cubical agda: init at version 0.2

laMudri commented:
Possibly some modifications will be needed when #76653 is merged. /cc @alexarice

alexarice commented:
This would need to be modified slightly, either this can be merged and I rebase on top of it or the other way round

alexarice commented:
Might be easier to merge this first

ryanorendorff commented:
Doesn't matter to me. I can also modify this if the other PR is merged first.

veprbl commented:
@ryanorendorff This will need to be adapted for the new Agda infrastructure. See manual https://github.com/NixOS/nixpkgs/blob/d30e2468e0c7875d3d4d47404f52647ccea76fcf/doc/languages-frameworks/agda.section.md

alexarice commented:
Note that v0.1 won't compile with agda 2.6.1 and there is not a more recent version

veprbl commented:
@alexarice This warrants getting more versions of Agda in nixpkgs?

alexarice commented:
Perhaps, I feel the libraries should release a version that works with the new agda version, as the master versions of these libraries work

ryanorendorff commented:
We could choose some development commit that passes their integration server for now. 

> On May 15, 2020, at 12:39, Alex Rice <notifications@github.com> wrote:
> 
> ﻿
> Perhaps, I feel the libraries should release a version that works with the new agda version, as the master versions of these libraries work
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

alexarice commented:
I feel this is the sensible option

ryanorendorff commented:
Ok I’ll find a good commit and update the PR when it builds correctly. 

> On May 15, 2020, at 12:47, Alex Rice <notifications@github.com> wrote:
> 
> ﻿
> I feel this is the sensible option
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Saizan commented:
I ended up tagging a recent master as the 0.2 release, it does work with 2.6.1:
https://github.com/agda/cubical/releases/tag/v0.2

alexarice commented:
@Saizan I tried out that commit and get the following error:
```
these derivations will be built:
  /nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv
building '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv'...
unpacking sources
unpacking source archive /nix/store/r5mlcnbhcp77y5wn7xxhggyi1wsaqnas-source
source root is source
patching sources
configuring
no configure script, doing nothing
building
runhaskell ./Everythings.hs gen-except Core Foundations Codata Experiments
Everythings.hs: ./Cubical/Data/DescendingList/Strict.agda: hGetContents: invalid argument (invalid byte sequence)
make: *** [GNUmakefile:30: gen-everythings] Error 1
builder for '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv' failed with exit code 2
error: build of '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv' failed
```

ryanorendorff commented:
I just got the same error. I'll push up what I have at the moment.

alexarice commented:
The problem doesn't happen outside a nix-build

Saizan commented:
I believe the script expects the locale to be UTF-8, is that something that can be set for the build?

alexarice commented:
I think it can

ryanorendorff committed:
Add ryanorendorff as a maintainer.

Saizan commented:
One can test it with `ghc -e "print System.IO.localeEncoding"`.

However `make all` is currently doing testing, it's not something useful for "installation"

alexarice commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ stdenv, mkDerivation, fetchFromGitHub, ghcWithPackages }:
```
```suggestion
{ lib, mkDerivation, fetchFromGitHub, ghcWithPackages }:
```

alexarice commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+  meta = with stdenv.lib; {
+    description =
+      "A cubical type theory library for use with the Agda compiler";
+    license = stdenv.lib.licenses.mit;
```
```suggestion
    license = licenses.mit;
```

ryanorendorff commented:
The output from the `localeEncoding` is `ASCII`. I'll try to change it to UTF-8.

alexarice commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+    description =
+      "A cubical type theory library for use with the Agda compiler";
+    license = stdenv.lib.licenses.mit;
+    platforms = stdenv.lib.platforms.unix;
```
```suggestion
    platforms = platforms.unix;
```

alexarice commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+      "A cubical type theory library for use with the Agda compiler";
+    license = stdenv.lib.licenses.mit;
+    platforms = stdenv.lib.platforms.unix;
+    maintainers = with maintainers; [ ryanorendorff ];
```
Can you add me as a maintainer

ryanorendorff replied:
Sure!

alexarice commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+    description =
+      "A cubical type theory library for use with the Agda compiler";
+    license = stdenv.lib.licenses.mit;
+    platforms = stdenv.lib.platforms.unix;
```
Missing a homepage field. I believe `homepage = src.meta.homepage` should work

ryanorendorff replied:
Added using the suggested statement.

ryanorendorff committed:
maintainers: add ryanorendorff githubId

ryanorendorff commented on maintainers/maintainer-list.nix:
```diff
@@ -6700,6 +6700,11 @@
     name = "Ryan Artecona";
   };
+  ryanorendorff = {
+    email = "12442942+ryanorendorff@users.noreply.github.com";
```
Added. Thanks!

@Saizan Do you want to be added as a maintainer?

Saizan replied:
I have never used nix myself, so it probably wouldn't help much for me to be a maintainer.

ryanorendorff replied:
Okie dokie. :+1:

ryanorendorff committed:
cubical agda: function call uses lib instead of stdenv

Co-authored-by: Alex Rice <alexrice999@hotmail.co.uk>

ryanorendorff committed:
cubical agda: function call uses lib instead of stdenv for meta

Co-authored-by: Alex Rice <alexrice999@hotmail.co.uk>

ryanorendorff committed:
cubical agda: function call uses lib instead of stdenv for platforms

Co-authored-by: Alex Rice <alexrice999@hotmail.co.uk>

ryanorendorff committed:
cubical agda: function call uses lib instead of stdenv for license

Co-authored-by: Alex Rice <alexrice999@hotmail.co.uk>

alexarice commented:
I think `LC_ALL = "en_US.UTF-8";` is what we are looking for with the locale

ryanorendorff committed:
cubical agda: add alexarice as maintainer

ryanorendorff commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+    make
+  '';
+
+  sourceDirectories = [ "." ];
```
Removed.

ryanorendorff committed:
cubical agda: remove unnecessary sourceDirectories

ryanorendorff commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+  # compiled through the make file they provide.
+  nativeBuildInputs = [ (ghcWithPackages (self: [ ])) ];
+  buildPhase = ''
+    patchShebangs
```
No I just added it temporarily while I was matching what the `standard-library` when first making the changes to the new agda setup.

Removed.

ryanorendorff committed:
cubical agda: remove unnecessary patchShebangs

ryanorendorff commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+  # The cubical library has several `Everything.agda` files, which are
+  # compiled through the make file they provide.
+  nativeBuildInputs = [ (ghcWithPackages (self: [ ])) ];
```
I agree. I had it this way while I was debugging. I'll remove this once the build is working with no packages.

ryanorendorff replied:
Replaced with just `ghc`.

ryanorendorff committed:
cubical agda: add homepage to meta

ryanorendorff commented:
> I think `LC_ALL = "en_US.UTF-8";` is what we are looking for with the locale

Hmm this leads to fun other errors:

```
runhaskell ./Everythings.hs gen-except Core Foundations Codata Experiments
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Everythings.hs: ./Cubical/Data/DescendingList/Strict.agda: hGetContents: invalid argument (invalid byte sequence)
make: *** [GNUmakefile:30: gen-everythings] Error 1
builder for '/nix/store/kh89hx2jpkz4kfy3rzwsdvvvpwdjhi37-cubical-0.2.drv' failed with exit code 2
```

alexarice commented:
Try to add `glibcLocales` to `buildInputs'?

ryanorendorff commented:
> Try to add `glibcLocales` to `buildInputs`? EDIT: Probably `nativeBuildInputs`

Yes that worked. Building now.

ryanorendorff committed:
cubical agda: replace ghcWithPackages with ghc

ryanorendorff committed:
cubical agda: set locale to UTF-8

Fixes the following build error.

```
building
runhaskell ./Everythings.hs gen-except Core Foundations Codata Experiments
Everythings.hs: ./Cubical/Data/DescendingList/Strict.agda: hGetContents: invalid argument (invalid byte sequence)
make: *** [GNUmakefile:30: gen-everythings] Error 1
```

ryanorendorff commented:
Builds correctly and I was able to type check some cubical code that I had.

ryanorendorff committed:
cubical agda: removed newline in agda-packages

alexarice reviewed (approved):
Nix-review passes
Package builds on nixos
Looks ready apart from rebase

alexarice commented:
@GrahamcOfBorg build agdaPackages.cubical

laMudri commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "07qlp2f189jvzbn3aqvpqk2zxpkmkxhhkjsn62iq436kxqj3z6c2";
+  };
+
+  LC_ALL = "en_US.UTF-8";
```
Does this not affect nearly every other Agda package? It feels like it should be the default.

alexarice replied:
Maybe though we haven't run into this problem before. Also this only seems to arise on the haskell building of Everything files and I don't think it is to do with `agda` I'm not sure it is correct to add it to the builder

laMudri replied:
Aah, okay, that explains a bit what's going on. I see stdlib's equivalent Haskell file makes sure everything is done in UTF-8, which is probably the right thing to do upstream.

alexarice commented:
Thanks @ryanorendorff , @Saizan , and @veprbl for merging 😄
